crates/block-producer/src/mempool/batch_graph.rs

```diff
@@ -0,0 +1,179 @@
+            for parent in &batch.parents {
+                // Parent could already be removed as part of this subgraph removal.
+                if let Some(parent) = self.nodes.get_mut(parent) {
+                    parent.children.remove(&batch_id);
+                }
+            }
+
+            to_process.extend(batch.children);
+            removed.push((batch_id, batch.transactions));
+        }
+
+        removed
+    }
+
+    /// Removes a set of batches from the graph without removing any descendents.
+    ///
+    /// This is intended to cull completed batches from stale blocks.
+    pub fn remove_stale(&mut self, batches: Vec<BatchJobId>) -> Vec<TransactionId> {
+        let mut transactions = Vec::new();
+
+        for batch in batches {
+            let node = self.nodes.remove(&batch).expect("Node must be in graph");
+            assert_eq!(node.status, Status::InBlock);
+
+            // Remove batch from graph. No need to update parents as they should be removed in this
+            // call as well.
+            for child in node.children {
+                // Its possible for the child to part of this same set of batches and therefore
+                // already removed.
+                if let Some(child) = self.nodes.get_mut(&child) {
+                    child.parents.remove(&batch);
+                }
+            }
+
+            transactions.extend_from_slice(&node.transactions);
+        }
+
+        transactions
+    }
+
+    /// Mark a batch as proven if it exists.
+    pub fn mark_proven(&mut self, id: BatchJobId) {
+        // Its possible for inflight batches to have been removed as part
+        // of another batches failure.
+        if let Some(node) = self.nodes.get_mut(&id) {
+            node.status = Status::Proven;
+            self.try_make_root(id);
+        }
+    }
+
+    pub fn pop_for_blocking(&mut self) -> Option<(BatchJobId, Vec<TransactionId>)> {
+        let batch_id = self.roots.pop_first()?;
+        let node = self.nodes.get_mut(&batch_id).expect("Root node must be in graph");
+        node.status = Status::InBlock;
+
+        // Work around multiple mutable borrows of self.
+        let transactions = node.transactions.clone();
+        let children = node.children.clone();
+
+        for child in children {
+            self.try_make_root(child);
+        }
+
+        Some((batch_id, transactions))
+    }
+
+    fn try_make_root(&mut self, id: BatchJobId) {
+        let node = self.nodes.get_mut(&id).expect("Node must be in graph");
+
+        for parent in node.parents.clone() {
+            let parent = self.nodes.get(&parent).expect("Parent must be in pool");
+
+            if parent.status != Status::InBlock {
+                return;
+            }
+        }
+        self.roots.insert(id);
+    }
+}
+
+#[derive(Clone, Debug)]
+struct Node {
+    status: Status,
+    transactions: Vec<TransactionId>,
+    parents: BTreeSet<BatchJobId>,
+    children: BTreeSet<BatchJobId>,
+}
```

Comment:
I will need to have 2 different variants of batches: `SelectedBatch` and `ProvenBatch`. The main difference between the two is that `ProvenBatch` would also eventually have a proof of correct construction attached to it (and there may be other differences as well). But I think we should make provisions for them now (even if the underlying structs are identical for now).

Maybe we can track the difference through the `Status` field - but we could have a proper enum too. Maybe something like:
```rust
struct Node {
    data: Batch,
    parents: BTreeSet<BatchJobId>,
    children: BTreeSet<BatchJobId>,
}

enum Batch {
    InFlight(SelectedBatch),
    Proven(ProventBatch),
    InBlock(ProvenBatch),
}
```